snapcraft/internal/repo/snaps.py

```diff
@@ -0,0 +1,196 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2017 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+import logging
+import sys
+import urllib
+from subprocess import check_call, check_output, CalledProcessError
+
+import requests_unixsocket
+from requests import exceptions
+
+from . import errors
+
+
+logger = logging.getLogger(__name__)
+_SNAPD_URI_ROOT_TEMPLATE = 'http+unix://%2Frun%2Fsnapd.socket/v2/{}'
+
+
+class SnapPackage:
+
+    @classmethod
+    def is_valid_snap(cls, snap):
+        return cls(snap).is_valid()
+
+    def __init__(self, snap):
+        """Lifecycle handler for a snap of the format <snap-name>/<channel>."""
+        self.name, self.channel = _get_parsed_snap(snap)
+        self._original_channel = self.channel
+        if not self.channel or self.channel == 'stable':
+            self.channel = 'latest/stable'
+
+        # This store information from a local request
+        self._local_snap_info = None
+        # And this stores information from a remote request
+        self._store_snap_info = None
+
+        self._is_installed = None
+        self._is_on_store = None
+
+    @property
+    def installed(self):
+        if self._is_installed is None:
+            self.get_local_snap_info()
+        return self._is_installed
```

Comment:
I think it's the property the one that should assign _is_installed, instead of doing it in get_local_snap_info

Reply:
Another question, can the installed status of the snap change after this property was called the first time? This returns the value of the first call, and never checks again. It feels weird, because the snap can be installed or uninstalled before the second call.

Reply:
it can change even while the call is taking place even if we iteratively call over and over again. It is just like query and install with `apt`